agenthub/codeact_agent/codeact_agent.py

```diff
@@ -191,48 +210,71 @@ def step(self, state: State) -> Action:
         if latest_user_message and latest_user_message.strip() == '/exit':
             return AgentFinishAction()
 
-        # prepare what we want to send to the LLM
-        messages: list[dict[str, str]] = self._get_messages(state)
-
-        response = self.llm.completion(
-            messages=messages,
-            stop=[
-                '</execute_ipython>',
-                '</execute_bash>',
-                '</execute_browse>',
-            ],
-            temperature=0.0,
-        )
+        response = None
+        # give it multiple chances to get a response
+        # if it fails, we'll try to condense memory
+        attempt = 0
+        while not response and attempt < self.llm.config.attempts_to_condense:
+            # prepare what we want to send to the LLM
+            messages: list[Message] = self._get_messages(state)
+            print('No of tokens, ' + str(self.llm.get_token_count(messages)) + '\n')
+            response = self.llm.completion(
+                messages=messages,
+                stop=[
+                    '</execute_ipython>',
+                    '</execute_bash>',
+                    '</execute_browse>',
+                ],
+                temperature=0.0,
+                condense=True,
+            )
+            attempt += 1
+
         return self.action_parser.parse(response)
 
-    def _get_messages(self, state: State) -> list[dict[str, str]]:
+    def search_memory(self, query: str) -> list[str]:
+        raise NotImplementedError('Implement this abstract method')
```

Comment:
We did? Oops, I'm slowly catching up with all changes lately. 😅 

Are we going to replace it with something? I think the lack of a recall action is an important reason why the memory work so far didn't perform so well in evals. (more below)

Reply:
Yep - the main reason is that "search_memory" function is never used by any agent in the hub and it doesn't make too much sense to me to put it as an agent's method (rather than a method of Runtime), since the idea is "Runtime" should be responsible for action execution, not agent.

This PR is one attempt to replace it. Another option would be have that as a optional agent skill. But let's see if this PR can solve this issue for now, and add that skill back when it is absolutely needed and we need to run eval to test them properly 🤔

Reply:
It was used by the monologue agent, and if I remember correctly, for a little time it was implemented also by the original SWE agent. So when that was gone, so was its usage.

You're right recalling stuff doesn't exactly belong to the agent... it didn't catch up, too, it was there in monologue from the start after all. 😅 I'm not sure it belongs to the runtime either, though, some actions don't. TBH the way I think of it, it belongs to long/short term memory. I see memory as a separate component that retrieves persistent stuff and injects or replaces stuff in the prompt on the fly, as needed and/or requested at every step. Like `history` started to do, where it replaces events with their summary or the other way around. That doesn't seem the same with "running runnable actions", it doesn't affect the agent's regular environment when executed, and for example it doesn't have anything to do with the sandboxing that we need for that environment, I don't think?

I found the relevant commits, thanks. I see where we're at. 🤔